src/source-maps.cc

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2017 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "source-maps.h"
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+
+#include "json/json.h"
+
+#define INDEX_NONE static_cast<size_t>(-1)
+
+#define INVALID()       \
+  do {                  \
+    if (fatal) abort(); \
+    return false;       \
+  } while (0);
+
+bool SourceMap::Validate(bool fatal) const {
+  for (size_t i = 0; i < segment_groups.size(); ++i) {
+    const auto& group = segment_groups[i];
+    if (i > 0 && group.generated_line <= segment_groups[i - 1].generated_line) {
```

Comment:
I've been writing this sort of thing as `CHECK_FOO(<cond>)` -- I find it more readable, but curious what you think.